pkg/scheduler/scheduler_test.go

```diff
@@ -1030,3 +1033,145 @@ func TestSchedulerBinding(t *testing.T) {
+	fakeWatch := watch.NewFake()
+	client.PrependWatchReactor("pods", ktesting.DefaultWatchReactor(fakeWatch, nil))
+	eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: client.EventsV1beta1().Events("")})
+
+	podsBound := make(chan *v1.Pod)
+	defer close(podsBound)
+
+	fns := []st.RegisterPluginFunc{
+		st.RegisterQueueSortPlugin(queuesort.Name, queuesort.New),
+		st.RegisterBindPlugin("fakeBinder", func(_ *runtime.Unknown, f framework.FrameworkHandle) (framework.Plugin, error) {
+			return newFakeBinder(fakeWatch, podsBound), nil
+		}),
+		st.RegisterPluginAsExtensions(podtopologyspread.Name, 1, podtopologyspread.New, "Filter", "PreFilter"),
+	}
+	emptySnapshot := internalcache.NewEmptySnapshot()
+	fwk, _ := st.NewFramework(
+		fns,
+		framework.WithClientSet(client),
+		framework.WithSnapshotSharedLister(emptySnapshot),
+	)
+	podQueue := internalqueue.NewSchedulingQueue(fwk)
+	algo := core.NewGenericScheduler(
+		cache,
+		podQueue,
+		emptySnapshot,
+		fwk,
+		[]core.SchedulerExtender{},
+		nil,
+		informerFactory.Core().V1().PersistentVolumeClaims().Lister(),
+		informerFactory.Policy().V1beta1().PodDisruptionBudgets().Lister(),
+		false,
+		schedulerapi.DefaultPercentageOfNodesToScore,
+		false,
+	)
+	sched := &Scheduler{
+		SchedulerCache:  cache,
+		Algorithm:       algo,
+		Framework:       fwk,
+		NextPod:         internalqueue.MakeNextPodFunc(podQueue),
+		Error:           MakeDefaultErrorFunc(client, podQueue, cache),
+		VolumeBinder:    volumebinder.NewFakeVolumeBinder(&volumescheduling.FakeVolumeBinderConfig{AllBound: true}),
+		Recorder:        eventBroadcaster.NewRecorder(scheme.Scheme, "scheduler"),
+		StopEverything:  ctx.Done(),
+		SchedulingQueue: podQueue,
+	}
+	podInformer := informerFactory.Core().V1().Pods()
+	AddAllEventHandlers(sched, "", informerFactory, podInformer)
+	informerFactory.Start(ctx.Done())
+	informerFactory.WaitForCacheSync(ctx.Done())
+
+	// Start the main scheduler goroutine.
+	go wait.UntilWithContext(ctx, sched.scheduleOne, 0)
+
+	p1 := st.MakePod().Name("p1").UID("p1").Label("foo", "").SpreadConstraint(
+		1, "node", v1.DoNotSchedule, st.MakeLabelSelector().Exists("foo").Obj(),
+	).Obj()
+	p2 := st.MakePod().Name("p2").UID("p2").Label("foo", "").SpreadConstraint(
+		1, "node", v1.DoNotSchedule, st.MakeLabelSelector().Exists("foo").Obj(),
+	).Obj()
+
+	fakeWatch.Add(p1)
+	fakeWatch.Add(p2)
+	for i := 0; i < 2; i++ {
+		select {
+		case p := <-podsBound:
+			if p.Name == p1.Name {
+				p1 = p
+			} else if p.Name == p2.Name {
+				p2 = p
+			}
+		case <-time.After(time.Second * 2):
+			t.Fatalf("timeout waiting for p1 and p2 to be scheduled.")
+		}
+	}
+	p1Copy := p1.DeepCopy()
+	now := metav1.Now()
+	p1Copy.DeletionTimestamp = &now
+	// Simulate a Pod deletion operation - either by user or controller manager.
+	fakeWatch.Modify(p1Copy)
+	// p3 is simulated as a replacement Pod created by controller manager.
+	time.Sleep(time.Millisecond * 200)
```

Comment:
Without this line, this UT may fail. It's because even if the AddEvent for p3 comes after UpdateEvent for p1, scheduleOne() (afterwards snapshot()) may happen before updatePodInCache()'s completion - as they're two goroutines. In that case, p3 can be either scheduled to node1 or node2, instead of being forced to scheduled to where p1 was scheduled.

This is not observed/reported in real usage yet, but it's technically possible.

Reply:
Is 200ms enough for p3's update? Will this cause a flake test?

Reply:
Usually, it's enough as it's a unit test env. But anyway, let me make it more robust. PTAL.